// angular-generic-table/generic-table/components/gt-custom-component-factory.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
var core_1 = require("@angular/core");
var GtCustomComponent = (function () {
    function GtCustomComponent() {
        this.redrawEvent = new core_1.EventEmitter();
    }
    GtCustomComponent.prototype.$redraw = function () {
        this.redrawEvent.emit({ row: this.row, column: this.column });
    };
    return GtCustomComponent;
}());
exports.GtCustomComponent = GtCustomComponent;
var GtCustomComponentFactory = (function () {
    function GtCustomComponentFactory() {
        this.redrawEvent = new core_1.EventEmitter();
    }
    GtCustomComponentFactory.prototype.instance = function (instance) {
        instance.row = this.row;
        instance.column = this.column;
        instance.redrawEvent.subscribe(this.redrawEvent);
    };
    return GtCustomComponentFactory;
}());
GtCustomComponentFactory.decorators = [
    { type: core_1.Component, args: [{
                selector: 'gt-custom-component-factory',
                template: "<div appComponentAnchor [ctor]=\"type\" \n                                     [injector]=\"injector\" \n                                     (instance)=\"instance($event)\"></div>"
            },] },
];
/** @nocollapse */
GtCustomComponentFactory.ctorParameters = function () { return []; };
GtCustomComponentFactory.propDecorators = {
    'type': [{ type: core_1.Input },],
    'injector': [{ type: core_1.Input },],
    'row': [{ type: core_1.Input },],
    'column': [{ type: core_1.Input },],
    'redrawEvent': [{ type: core_1.Output },],
};
exports.GtCustomComponentFactory = GtCustomComponentFactory;
