// @angular-generic-table/core/esm2015/directives/component-anchor.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ComponentFactoryResolver, Directive, EventEmitter, Injector, Input, Output, Type, ViewContainerRef } from '@angular/core';
/**
 * @template C
 */
export class ComponentAnchorDirective {
    /**
     * @param {?} componentFactoryResolver
     * @param {?} viewContainer
     */
    constructor(componentFactoryResolver, viewContainer) {
        this.componentFactoryResolver = componentFactoryResolver;
        this.viewContainer = viewContainer;
        this.instance = new EventEmitter();
    }
    /**
     * @return {?}
     */
    ngOnInit() {
        const /** @type {?} */ factory = this.componentFactoryResolver.resolveComponentFactory(this.ctor);
        const /** @type {?} */ component = this.viewContainer.createComponent(factory, 0, this.injector);
        this.instance.emit(component.instance);
    }
}
ComponentAnchorDirective.decorators = [
    { type: Directive, args: [{
                selector: '[gtComponentAnchor]'
            },] },
];
/** @nocollapse */
ComponentAnchorDirective.ctorParameters = () => [
    { type: ComponentFactoryResolver, },
    { type: ViewContainerRef, },
];
ComponentAnchorDirective.propDecorators = {
    "ctor": [{ type: Input },],
    "injector": [{ type: Input },],
    "instance": [{ type: Output },],
};
function ComponentAnchorDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    ComponentAnchorDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    ComponentAnchorDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    ComponentAnchorDirective.propDecorators;
    /** @type {?} */
    ComponentAnchorDirective.prototype.ctor;
    /** @type {?} */
    ComponentAnchorDirective.prototype.injector;
    /** @type {?} */
    ComponentAnchorDirective.prototype.instance;
    /** @type {?} */
    ComponentAnchorDirective.prototype.componentFactoryResolver;
    /** @type {?} */
    ComponentAnchorDirective.prototype.viewContainer;
}
