// @angular-generic-table/core/esm5/directives/gt-column-width.directive.js
/**
 * @fileoverview added by tsickle
 * @suppress {checkTypes} checked by tsc
 */
import { ChangeDetectorRef, Directive, ElementRef, HostListener, Input } from '@angular/core';
var GtColumnWidthDirective = /** @class */ (function () {
    function GtColumnWidthDirective(hostElement, cdRef) {
        this.hostElement = hostElement;
        this.cdRef = cdRef;
    }
    /**
     * @param {?} $event
     * @return {?}
     */
    GtColumnWidthDirective.prototype.onResize = /**
     * @param {?} $event
     * @return {?}
     */
    function ($event) {
        this.checkSize();
    };
    /**
     * @return {?}
     */
    GtColumnWidthDirective.prototype.ngOnInit = /**
     * @return {?}
     */
    function () {
        this.checkSize();
    };
    /**
     * @return {?}
     */
    GtColumnWidthDirective.prototype.checkSize = /**
     * @return {?}
     */
    function () {
        this.widths[this.objectKey] = window
            .getComputedStyle(this.hostElement.nativeElement, null)
            .getPropertyValue('width');
        this.cdRef.detectChanges();
    };
    GtColumnWidthDirective.decorators = [
        { type: Directive, args: [{
                    selector: '[gtColumnWidth]'
                },] },
    ];
    /** @nocollapse */
    GtColumnWidthDirective.ctorParameters = function () { return [
        { type: ElementRef, },
        { type: ChangeDetectorRef, },
    ]; };
    GtColumnWidthDirective.propDecorators = {
        "objectKey": [{ type: Input },],
        "widths": [{ type: Input },],
        "onResize": [{ type: HostListener, args: ['window:resize', ['$event'],] },],
    };
    return GtColumnWidthDirective;
}());
export { GtColumnWidthDirective };
function GtColumnWidthDirective_tsickle_Closure_declarations() {
    /** @type {!Array<{type: !Function, args: (undefined|!Array<?>)}>} */
    GtColumnWidthDirective.decorators;
    /**
     * @nocollapse
     * @type {function(): !Array<(null|{type: ?, decorators: (undefined|!Array<{type: !Function, args: (undefined|!Array<?>)}>)})>}
     */
    GtColumnWidthDirective.ctorParameters;
    /** @type {!Object<string,!Array<{type: !Function, args: (undefined|!Array<?>)}>>} */
    GtColumnWidthDirective.propDecorators;
    /** @type {?} */
    GtColumnWidthDirective.prototype.objectKey;
    /** @type {?} */
    GtColumnWidthDirective.prototype.widths;
    /** @type {?} */
    GtColumnWidthDirective.prototype.hostElement;
    /** @type {?} */
    GtColumnWidthDirective.prototype.cdRef;
}
